hare/algorithm.go

```diff
@@ -0,0 +1,240 @@
+package hare
+
+import (
+	"errors"
+	"github.com/gogo/protobuf/proto"
+	"github.com/spacemeshos/go-spacemesh/hare/pb"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"sync"
+	"time"
+)
+
+const ProtoName = "HARE_PROTOCOL"
+const RoundDuration = time.Second * time.Duration(15)
+const InitialKnowledgeSize = 1000
+
+type Byteable interface {
+	Bytes() []byte
+}
+
+type NetworkService interface {
+	RegisterProtocol(protocol string) chan service.Message
+	Broadcast(protocol string, payload []byte) error
+}
+
+type State struct {
+	k    uint32      // the iteration number
+	ki   uint32      // ?
+	s    Set         // the set of blocks
+	cert Certificate // the certificate
+}
+
+type ConsensusProcess struct {
+	State
+	Closer // the consensus is closeable
+	pubKey      PubKey
+	layerId     LayerId
+	oracle      Rolacle // roles oracle
+	signing     Signing
+	network     NetworkService
+	startTime   time.Time // TODO: needed?
+	inbox       chan *pb.HareMessage
+	knowledge   []*pb.HareMessage
+	isProcessed map[uint32]bool // TODO: could be empty struct
+	mutex       sync.Mutex
+	// TODO: add knowledge of notify messages. What is the life-cycle of such messages? persistent according to Julian
+}
+
+func (proc *ConsensusProcess) NewConsensusProcess(layer LayerId, s Set, oracle Rolacle, signing Signing, p2p NetworkService, broker *Broker) {
+	proc.State = State{0, 0, s, Certificate{}}
+	proc.Closer = NewCloser()
+	proc.layerId = layer
+	proc.oracle = oracle
+	proc.signing = signing
+	proc.network = p2p
+	proc.inbox = broker.CreateInbox(&layer)
+	proc.knowledge = make([]*pb.HareMessage, 0)
+	proc.isProcessed = make(map[uint32]bool)
+}
+
+func (proc *ConsensusProcess) Start() {
+	if !proc.startTime.IsZero() { // called twice on same instance
+		log.Error("ConsensusProcess has already been started.")
+		return
+	}
+
+	proc.startTime = time.Now()
+
+	// send first status message
+	m, err := proc.buildStatusMessage()
+	if err == nil {
+		proc.network.Broadcast(ProtoName, m)
```

Comment:
is broadcast blocking? if so, consider moving it to a go routine

Reply:
Non blocking